### ISSUE TITLE
generateThreadConfig() Auto generated configurations Improvement request

### ISSUE THREAD

**Issue by Sweepi** [fireice-uk/xmr-stak#2057]
(I skipped the issue template, since I do not think that it fits the nature of this request)
(words with **emphasis** are variable names used in the code)

Hi,

I am wondering if and how the auto config generation can be improved.

Except*1 for the special handling of AMD Vega cards, **maxThreads** in generateThreadConfig()  [in autoAdjust.hpp] is always set to 1000.
*2 **maxIntensity** is set to "(Available VRAM/Scratchpad size)"
**possibleIntensity** is then set to the smaller value of **maxThreads** and **maxIntensity**.
In most cases, 1000 is smaller than (Available VRAM/Scratchpad size), therefore **possibleIntensity**  is set to 1000 in most cases.
generateThreadConfig()  then makes sure that **intensity** is the biggest multiple of (8 * **ctx.computeUnits**) which is still <= **possibleIntensity**.

Example: Fury X:
64 CUs
4 GiB VRAM
**possibleIntensity** = min (1000, (4 GiB - 128 MiB) / (2 MiB + 224))
**possibleIntensity** = min (1000,  ~2000)
**possibleIntensity** = 1000
64 CU * 8 = 512
The largest multiple of 512 which is <= 1000 is 512.

Therefore:
**intensity** = 512

512 works, but is not nearly optimal on a Fury X.
The hashrate _doubles_ when using an **intensity** of 1280.

There is an explanation for the magic value 1000, it says the following:
`			/* 1000 is a magic selected limit, the reason is that more than 2GiB memory
			 * sowing down the memory performance because of TLB cache misses
			 */
			size_t maxThreads = 1000u;`

Can you think of an algorithm to calculate  **maxThreads** more tailored to the specific card? What does it depend on? My uneducated guess: GCN version, number of CUs, anything else?

*1 there is also special handling for NVIDIA P100 and V100 devices, which I omitted for simplicity.
*2 actually: **maxIntensity** = (VRAM - 128 MiB) / (Scratchpad size  + 224)

**Comment by Spudz76** [comment]
Management of all the various autoconfigure math for various GPU types would almost be a project in itself.
Autoconfig is just meant as a place to start finding out what works best on your rig/driver/kernel/winver, by hand.

**Comment by Sweepi** [comment]
Well, for me this is about learning how it could be done better and aiming for to be able contribute so that it will be better in the future.

**Comment by psychocrypt** [comment]
yes it is very hard to have good fixed rules. For vega ee have also an
extra path to have a higher intensity than 1000.

**Comment by Spudz76** [comment]
Really the only way this could work is an autotester that literally benchmarks all of the possibilities for you (on say 1024 hashes per config, quick but enough to catch a sense of what's fastest).
nVidia backend would have to be upgraded to use NVRTC or it would have to re-init the GPUs completely in between minibench loops.  NVRTC should be done anyways to make that backend more aligned with how the OpenCL backend works (compile at runtime, can use `#DEFINE` to drop unused code, compiles the shader code with whatever driver the user is running...

That or something mentioned before was a phonehome feature that reports the backend configs and resulting rates to a database somewhere.  Humans reporting their own results is too inaccurate, it must be internal to the app to be accurate at all.  Then when end-user finds the best settings for their GPU we have it and can apply that to the autoconfig.  More database driven than human intervention required.

All other methods seem like they will just become less correct over time, versus more correct.
Most autoconfig code currently based on "guessing" and maybe a few hotfixes for where it was extra wrong on weird edge cases (GPUs like five users own).  Then Vega is too new and the driver(s) have still been changing in somewhat major ways pretty often so there is no autoconfig for those despite they would be faster then (it would have to probe driver version, and switch logic based on that, making the autoconf a gigantic messy tree).
